rainier-core/src/main/scala/rainier/compute/Real.scala

```diff
@@ -69,6 +69,21 @@ object Real {
     loop(real, Set.empty)
   }
 
+  def substituteVariable(x: Real, f: Variable => Real): Real = x match {
```

Comment:
This is great. You should memoize the transformation to avoid wasted work and so that nodes that start out identical end up identical.

Reply:
What's the reason for taking `Variable => Real` instead of `Map[Variable,Real]` which is what you are actually providing - just that the former seems more general? I think I'd lean somewhat towards the concreteness/explicitness of the latter.